rascaline/src/calculators/composition.rs

```diff
@@ -0,0 +1,169 @@
+use std::sync::Arc;
+
+use equistore::{Labels, LabelsBuilder, TensorMap};
+
+use crate::{Error, System};
+
+use super::CalculatorBase;
+use crate::labels::{SpeciesFilter, SamplesBuilder};
+use crate::labels::SamplesPerAtom;
+use crate::labels::{CenterSpeciesKeys, KeysBuilder};
+
+#[derive(Debug, Clone)]
+#[derive(serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+/// An composition calculator for obtaining the stoichiometric information.
+pub struct Composition {}
+
+impl CalculatorBase for Composition {
+    fn name(&self) -> String {
+        return "atom centered composition features".into();
+    }
+
+    fn parameters(&self) -> String {
+        return serde_json::to_string(self).expect("failed to serialize to JSON");
+    }
+
+    fn keys(&self, systems: &mut [Box<dyn System>]) -> Result<Labels, Error> {
+        return CenterSpeciesKeys.keys(systems);
+    }
+
+    fn samples_names(&self) -> Vec<&str> {
+        return SamplesPerAtom::samples_names();
+    }
+
+    fn samples(&self, keys: &Labels, systems: &mut [Box<dyn System>]) -> Result<Vec<Arc<Labels>>, Error> {
+        assert_eq!(keys.names(), ["species_center"]);
+        let mut samples = Vec::new();
+        for [species_center] in keys.iter_fixed_size() {
+            let builder = SamplesPerAtom {
+                species_center: SpeciesFilter::Single(species_center.i32())
+            };
+    
+            samples.push(builder.samples(systems)?);
+        }
+
+        return Ok(samples);
+    }
+
+    fn supports_gradient(&self, parameter: &str) -> bool {
+        match parameter {
+            "positions" => true,
+            "cell" => true,
+            _ => false,
+        }
+    }
+
+    fn positions_gradient_samples(&self, keys: &Labels, samples: &[Arc<Labels>], systems: &mut [Box<dyn System>]) -> Result<Vec<Arc<Labels>>, Error> {
+        debug_assert_eq!(keys.count(), samples.len());
+        let mut gradient_samples = Vec::new();
+        for ([species_center], samples) in keys.iter_fixed_size().zip(samples) {
+            let builder = SamplesPerAtom {
+                species_center: SpeciesFilter::Single(species_center.i32())
+            };
+
+            gradient_samples.push(builder.gradients_for(systems, samples)?);
+        }
+
+        return Ok(gradient_samples);
+    }
+
+    fn components(&self, keys: &Labels) -> Vec<Vec<Arc<Labels>>> {
+        return vec![Vec::new(); keys.count()];
+    }
+
+    fn properties_names(&self) -> Vec<&str> {
+        return vec!["count"];
```

Comment:
I think there might be an issue here. A single "count" property label does not make this kind of descriptor easily summable over different atoms in the same structure. To do that, in my opinion, we would instead need a "species" property label, which has dimension n_species (i.e., the properties dimension is n_species long) and which has value 1 for the species of the central atom and zero otherwise. This would make it summable across atoms in the same structure but having a different central atom species.